src/main/java/com/couchbase/client/dcp/state/SessionState.java

```diff
@@ -164,12 +165,12 @@ public void rollbackToPosition(short partition, long seqno) {
         ps.setStartSeqno(seqno);
         ps.setSnapshotStartSeqno(seqno);
         ps.setSnapshotEndSeqno(seqno);
-        Iterator<FailoverLogEntry> flog = ps.getFailoverLog().iterator();
-        while (flog.hasNext()) {
-            FailoverLogEntry entry = flog.next();
+        List<FailoverLogEntry> flog = ps.getFailoverLog();
+        for (int i = 0; i < flog.size(); i ++) {
```

Comment:
You should invert the cycle:

```
for (int i = flog.size() - 1; i > 0; i--) {
```

otherwise you may skip processing some items, i.e. while i is 2 for example and supposing we should remove this enrty, when we increase i to 3 in the next cycle, flog.get(3) will give you back the flog.get(4) of the original size!

Reply:
@themnd Thanks!